src/model/ImageDocument.hpp

```diff
@@ -70,39 +74,59 @@ class ImageDocument : public QObject
   bool isComparison() const { return (bool)comparison_; }
   boost::optional<ImageComparison> const& comparison() const { return comparison_; }
 
+  enum ErrorCategory { E_Image, E_Comparsion, E_Generic };
+  void setError(QString const& errorText, ErrorCategory category);
+
   void setPosition(QPointF pos);
   void move(QPointF offset);
   void setScale(qreal scale);
   void setBrightness(qreal brightness);
   void setGamma(qreal gamma);
   void setCurrentPixel(QPoint index);
 
-  Q_INVOKABLE bool store(QUrl const& url);
+  Q_INVOKABLE void resetError();
+  Q_INVOKABLE void store(QUrl const& url);
 
 signals:
+  void busyChanged();
+  void errorTextChanged();
   void propertyChanged();
   void positionChanged();
   void scaleChanged();
   void brightnessChanged();
   void gammaChanged();
   void pixelPositionChanged();
   void pixelValueChanged();
+  void fileTypeChanged();
 
 private:
-
-  bool check(Result<bool> const&);
-  void setError(QString const& message);
+  typedef std::shared_ptr<Result<Image>> LoadResult;
+  void load(QString const& path, QFutureWatcher<LoadResult>* watcher);
+  void loadFinished(QFutureWatcher<LoadResult>* watcher, QUrl const& url, bool comparison);
+
+  template<class T>
+  bool check(Result<T> const& result, ErrorCategory category, QString const& prefix = "") {
+    if (!result && !result.error().empty()) {
```

Comment:
`!result` should imply `!result.error().empty()` already, I guess there could be an error without an empty message but then it would still be good if setError gets called I think